File: app/core/game_engine/engine.py
# CORE_CANDIDATE
"""Game engine — manages the full lifecycle of a best-of-3 match.

Orchestrates turn resolution, round tracking, and match completion.
Pure state machine with no I/O dependencies. Reusable for any
simultaneous-action turn-based game.
"""

from app.core.game_engine.outcome_matrix import (
    resolve_turn as _resolve_turn,
    validate_action as _validate_action,
)
from app.core.game_engine.types import (
    Action,
    GameState,
    MatchResult,
    MatchStatus,
    MatchType,
    ROUNDS_TO_WIN,
    RoundResult,
    RoundState,
    RoundWinner,
    TURN_LIMIT,
    TurnOutcome,
    TurnResult,
)


class GameEngine:
    """Stateless engine that advances a GameState through turns, rounds, and match.

    All methods take a GameState and return a new/modified state.
    No side effects, no I/O — suitable for both AI and PvP modes.
    """

    def validate_action(self, action: Action, player_ki: int) -> bool:
        """Check whether a player can afford an action.

        Args:
            action: The action the player wants to take.
            player_ki: The player's current ki.

        Returns:
            True if the player has enough ki.
        """
        return _validate_action(action, player_ki)

    def start_match(self, match_type: MatchType) -> GameState:
        """Create a new match and start the first round.

        Args:
            match_type: Type of match (AI difficulty or PvP).

        Returns:
            GameState with round 1 initialized.
        """
        state = GameState(
            match_type=match_type,
            status=MatchStatus.IN_PROGRESS,
        )
        return self._start_new_round(state)

    def submit_turn(
        self,
        state: GameState,
        p1_action: Action,
        p2_action: Action,
    ) -> tuple[GameState, TurnResult, RoundResult | None, MatchResult | None]:
        """Submit both players' actions and resolve the turn.

        Validates actions, resolves the turn, checks for round/match end.

        Args:
            state: Current game state.
            p1_action: Player 1's chosen action.
            p2_action: Player 2's chosen action.

        Returns:
            Tuple of (updated GameState, TurnResult, optional RoundResult,
            optional MatchResult).

        Raises:
            ValueError: If match is not in progress or no active round.
            ValueError: If a player cannot afford their chosen action.
        """
        if state.status != MatchStatus.IN_PROGRESS:
            raise ValueError(f"Match is not in progress (status={state.status})")

        if state.current_round is None:
            raise ValueError("No active round")

        current_round = state.current_round

        if not self.validate_action(p1_action, current_round.p1_ki):
            raise ValueError(
                f"P1 cannot afford {p1_action.value} (ki={current_round.p1_ki})"
            )
        if not self.validate_action(p2_action, current_round.p2_ki):
            raise ValueError(
                f"P2 cannot afford {p2_action.value} (ki={current_round.p2_ki})"
            )

        turn_number = current_round.turn_number + 1
        turn_result = _resolve_turn(
            turn_number=turn_number,
            p1_action=p1_action,
            p2_action=p2_action,
            p1_ki=current_round.p1_ki,
            p2_ki=current_round.p2_ki,
        )

        # Update round state
        current_round.turn_number = turn_number
        current_round.p1_ki = turn_result.p1_ki_after
        current_round.p2_ki = turn_result.p2_ki_after
        current_round.turn_history.append(turn_result)

        # Check round end
        round_result = self._check_round_end(turn_result, current_round)
        match_result = None

        if round_result is not None:
            if round_result.winner == RoundWinner.P1:
                state.rounds_won_p1 += 1
            elif round_result.winner == RoundWinner.P2:
                state.rounds_won_p2 += 1

            state.round_results.append(round_result)
            state.current_round = None

            match_result = self._check_match_end(state)

            if match_result is not None:
                state.status = MatchStatus.COMPLETED
            else:
                state = self._start_new_round(state)

        return state, turn_result, round_result, match_result

    def forfeit(
        self,
        state: GameState,
        forfeiting_player: RoundWinner,
    ) -> tuple[GameState, MatchResult]:
        """Handle a player forfeiting the match.

        Args:
            state: Current game state.
            forfeiting_player: Which player is forfeiting (P1 or P2).

        Returns:
            Tuple of (updated GameState, MatchResult).
        """
        state.status = MatchStatus.ABANDONED
        winner = (
            RoundWinner.P2 if forfeiting_player == RoundWinner.P1 else RoundWinner.P1
        )

        total_turns = sum(r.total_turns for r in state.round_results)
        if state.current_round:
            total_turns += state.current_round.turn_number

        match_result = MatchResult(
            game_id=state.game_id,
            winner=winner,
            rounds_won_p1=state.rounds_won_p1,
            rounds_won_p2=state.rounds_won_p2,
            round_results=state.round_results,
            total_turns=total_turns,
        )
        return state, match_result

    def _start_new_round(self, state: GameState) -> GameState:
        """Initialize a new round within the match."""
        round_number = len(state.round_results) + 1
        state.current_round = RoundState(
            round_number=round_number,
            p1_ki=0,
            p2_ki=0,
            turn_number=0,
        )
        return state

    def _check_round_end(
        self,
        turn_result: TurnResult,
        round_state: RoundState,
    ) -> RoundResult | None:
        """Check if the current round has ended.

        A round ends when:
        1. A player wins (P1_WINS_ROUND or P2_WINS_ROUND outcome).
        2. Turn limit (20) reached — player with more ki wins; tie → DRAW.
        """
        winner: RoundWinner | None = None

        if turn_result.outcome == TurnOutcome.P1_WINS_ROUND:
            winner = RoundWinner.P1
        elif turn_result.outcome == TurnOutcome.P2_WINS_ROUND:
            winner = RoundWinner.P2
        elif round_state.turn_number >= TURN_LIMIT:
            if round_state.p1_ki > round_state.p2_ki:
                winner = RoundWinner.P1
            elif round_state.p2_ki > round_state.p1_ki:
                winner = RoundWinner.P2
            else:
                winner = RoundWinner.DRAW

        if winner is None:
            return None

        return RoundResult(
            round_number=round_state.round_number,
            winner=winner,
            total_turns=round_state.turn_number,
            final_p1_ki=round_state.p1_ki,
            final_p2_ki=round_state.p2_ki,
        )

    def _check_match_end(self, state: GameState) -> MatchResult | None:
        """Check if the match has ended.

        Match ends when a player reaches ROUNDS_TO_WIN (2) round wins,
        or all 3 rounds played with no winner (1-1 + draw → draw match).
        """
        total_turns = sum(r.total_turns for r in state.round_results)

        if state.rounds_won_p1 >= ROUNDS_TO_WIN:
            return MatchResult(
                game_id=state.game_id,
                winner=RoundWinner.P1,
                rounds_won_p1=state.rounds_won_p1,
                rounds_won_p2=state.rounds_won_p2,
                round_results=state.round_results,
                total_turns=total_turns,
            )

        if state.rounds_won_p2 >= ROUNDS_TO_WIN:
            return MatchResult(
                game_id=state.game_id,
                winner=RoundWinner.P2,
                rounds_won_p1=state.rounds_won_p1,
                rounds_won_p2=state.rounds_won_p2,
                round_results=state.round_results,
                total_turns=total_turns,
            )

        # 3 rounds played, no one has 2 wins → draw
        if len(state.round_results) >= 3:
            return MatchResult(
                game_id=state.game_id,
                winner=RoundWinner.DRAW,
                rounds_won_p1=state.rounds_won_p1,
                rounds_won_p2=state.rounds_won_p2,
                round_results=state.round_results,
                total_turns=total_turns,
            )

        return None

File: go-server/engine.go
package main

// Pure game engine — ports outcome_matrix.py + engine.py.
//
// No I/O, no globals. Every function takes the GameState (or pieces of
// it) and returns either a new value or applies mutation locally.
// Matches the Python contract byte-for-byte so AI-mode replays and PvP
// histories stay portable.

import (
	"fmt"

	"github.com/google/uuid"
)

// outcomeMatrix[p1][p2] = outcome from P1's perspective. Mirror of
// _MATRIX in app/core/game_engine/outcome_matrix.py.
var outcomeMatrix = map[Action]map[Action]TurnOutcome{
	ActionCharge: {
		ActionCharge:     OutcomeNeutral,
		ActionBlock:      OutcomeNeutral,
		ActionAttack:     OutcomeP2WinsRound,
		ActionEnergyWave: OutcomeP2WinsRound,
		ActionTeleport:   OutcomeNeutral,
	},
	ActionBlock: {
		ActionCharge:     OutcomeNeutral,
		ActionBlock:      OutcomeNeutral,
		ActionAttack:     OutcomeBlocked,
		ActionEnergyWave: OutcomeP2WinsRound,
		ActionTeleport:   OutcomeNeutral,
	},
	ActionAttack: {
		ActionCharge:     OutcomeP1WinsRound,
		ActionBlock:      OutcomeBlocked,
		ActionAttack:     OutcomeClash,
		ActionEnergyWave: OutcomeP2WinsRound,
		ActionTeleport:   OutcomeDodged,
	},
	ActionEnergyWave: {
		ActionCharge:     OutcomeP1WinsRound,
		ActionBlock:      OutcomeP1WinsRound,
		ActionAttack:     OutcomeP1WinsRound,
		ActionEnergyWave: OutcomeClash,
		ActionTeleport:   OutcomeDodged,
	},
	ActionTeleport: {
		ActionCharge:     OutcomeNeutral,
		ActionBlock:      OutcomeNeutral,
		ActionAttack:     OutcomeDodged,
		ActionEnergyWave: OutcomeDodged,
		ActionTeleport:   OutcomeNeutral,
	},
}

func validateAction(action Action, ki int) bool {
	return ki >= actionKiCost[action]
}

func calcKiAfter(action Action, currentKi int) int {
	ki := currentKi - actionKiCost[action] + actionKiGain[action]
	if ki < 0 {
		ki = 0
	}
	if ki > KiCap {
		ki = KiCap
	}
	return ki
}

func resolveTurn(turnNumber int, p1, p2 Action, p1Ki, p2Ki int) TurnResult {
	outcome := outcomeMatrix[p1][p2]
	return TurnResult{
		TurnNumber: turnNumber,
		P1Action:   p1,
		P2Action:   p2,
		Outcome:    outcome,
		P1KiBefore: p1Ki,
		P2KiBefore: p2Ki,
		P1KiAfter:  calcKiAfter(p1, p1Ki),
		P2KiAfter:  calcKiAfter(p2, p2Ki),
	}
}

// startNewRound seeds round N+1 inside the GameState.
func startNewRound(gs *GameState) {
	gs.CurrentRound = &RoundState{
		RoundNumber: len(gs.RoundResults) + 1,
		P1Ki:        0,
		P2Ki:        0,
		TurnNumber:  0,
		TurnHistory: []TurnResult{},
	}
}

// submitTurn applies both actions to the game state, advances rounds/match.
// Returns the turn result + optional round/match result. Mirrors
// engine.py GameEngine.submit_turn.
func submitTurn(gs *GameState, p1, p2 Action) (TurnResult, *RoundResult, *MatchResult, error) {
	if gs.Status != MatchInProgress {
		return TurnResult{}, nil, nil, fmt.Errorf("match not in progress (status=%s)", gs.Status)
	}
	if gs.CurrentRound == nil {
		return TurnResult{}, nil, nil, fmt.Errorf("no active round")
	}
	cr := gs.CurrentRound
	if !validateAction(p1, cr.P1Ki) {
		return TurnResult{}, nil, nil, fmt.Errorf("p1 cannot afford %s (ki=%d)", p1, cr.P1Ki)
	}
	if !validateAction(p2, cr.P2Ki) {
		return TurnResult{}, nil, nil, fmt.Errorf("p2 cannot afford %s (ki=%d)", p2, cr.P2Ki)
	}

	turnNum := cr.TurnNumber + 1
	tr := resolveTurn(turnNum, p1, p2, cr.P1Ki, cr.P2Ki)

	cr.TurnNumber = turnNum
	cr.P1Ki = tr.P1KiAfter
	cr.P2Ki = tr.P2KiAfter
	cr.TurnHistory = append(cr.TurnHistory, tr)

	rr := checkRoundEnd(tr, cr)
	var mr *MatchResult

	if rr != nil {
		switch rr.Winner {
		case WinnerP1:
			gs.RoundsWonP1++
		case WinnerP2:
			gs.RoundsWonP2++
		}
		gs.RoundResults = append(gs.RoundResults, *rr)
		gs.CurrentRound = nil

		mr = checkMatchEnd(gs)
		if mr != nil {
			gs.Status = MatchCompleted
		} else {
			startNewRound(gs)
		}
	}

	return tr, rr, mr, nil
}

func checkRoundEnd(tr TurnResult, rs *RoundState) *RoundResult {
	var winner RoundWinner
	switch tr.Outcome {
	case OutcomeP1WinsRound:
		winner = WinnerP1
	case OutcomeP2WinsRound:
		winner = WinnerP2
	default:
		if rs.TurnNumber >= TurnLimit {
			switch {
			case rs.P1Ki > rs.P2Ki:
				winner = WinnerP1
			case rs.P2Ki > rs.P1Ki:
				winner = WinnerP2
			default:
				winner = WinnerDraw
			}
		}
	}
	if winner == "" {
		return nil
	}
	return &RoundResult{
		RoundNumber: rs.RoundNumber,
		Winner:      winner,
		TotalTurns:  rs.TurnNumber,
		FinalP1Ki:   rs.P1Ki,
		FinalP2Ki:   rs.P2Ki,
	}
}

func checkMatchEnd(gs *GameState) *MatchResult {
	totalTurns := 0
	for _, r := range gs.RoundResults {
		totalTurns += r.TotalTurns
	}
	if gs.RoundsWonP1 >= RoundsToWin {
		return &MatchResult{
			GameID:       gs.GameID,
			Winner:       WinnerP1,
			RoundsWonP1:  gs.RoundsWonP1,
			RoundsWonP2:  gs.RoundsWonP2,
			RoundResults: gs.RoundResults,
			TotalTurns:   totalTurns,
		}
	}
	if gs.RoundsWonP2 >= RoundsToWin {
		return &MatchResult{
			GameID:       gs.GameID,
			Winner:       WinnerP2,
			RoundsWonP1:  gs.RoundsWonP1,
			RoundsWonP2:  gs.RoundsWonP2,
			RoundResults: gs.RoundResults,
			TotalTurns:   totalTurns,
		}
	}
	if len(gs.RoundResults) >= 3 {
		return &MatchResult{
			GameID:       gs.GameID,
			Winner:       WinnerDraw,
			RoundsWonP1:  gs.RoundsWonP1,
			RoundsWonP2:  gs.RoundsWonP2,
			RoundResults: gs.RoundResults,
			TotalTurns:   totalTurns,
		}
	}
	return nil
}

// newGameID returns a v4 UUID string — only used when this server creates
// a match directly. (In production we read game_ids issued by the Python
// rooms endpoint via Redis; this helper is for tests / standalone runs.)
func newGameID() string {
	return uuid.NewString()
}
